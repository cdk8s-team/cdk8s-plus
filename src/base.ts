import { ApiObjectMetadata, ApiObject, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiResource, IApiEndpoint } from './api-resource.generated';
import * as r from './role';
import * as rb from './role-binding';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  readonly apiVersion: string;

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io")
   */
  readonly apiGroup: string;

  /**
   * The object kind (e.g. "Deployment").
   */
  readonly kind: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource, IApiResource, IApiEndpoint {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: ApiObject;

  public get metadata(): ApiObjectMetadataDefinition {
    return this.apiObject.metadata;
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  public get apiVersion(): string {
    return this.apiObject.apiVersion;
  }

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io").
   */
  public get apiGroup(): string {
    return this.apiObject.apiGroup;
  }

  /**
   * The object kind (e.g. "Deployment").
   */
  public get kind(): string {
    return this.apiObject.kind;
  }

  public get resourceName(): string | undefined {
    return this.name;
  }

  public readonly abstract resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }

  /**
   * Grants the list of subjects permissions to read this resource.
   */
  public grantRead(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowRead(this);
    return role.bind(...subjects);
  }

  /**
   * Grants the list of subjects permissions to read and write this resource.
   */
  public grantReadWrite(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowReadWrite(this);
    return role.bind(...subjects);
  }

}