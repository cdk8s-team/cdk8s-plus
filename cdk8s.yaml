output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
