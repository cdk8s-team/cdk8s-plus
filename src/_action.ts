import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}