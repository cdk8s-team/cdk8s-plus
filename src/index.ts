export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
