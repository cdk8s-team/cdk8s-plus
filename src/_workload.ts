import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}