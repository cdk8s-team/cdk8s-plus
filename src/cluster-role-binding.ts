import { ApiObject, Lazy } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as clusterrole from './cluster-role';
import * as k8s from './imports/k8s';
import * as rb from './role-binding';
import { filterUndefined } from './utils';

/**
 * Properties for `ClusterRoleBinding`.
 */
export interface ClusterRoleBindingProps extends ResourceProps {
  /**
   * The role to bind to.
   */
  readonly role: clusterrole.IClusterRole;
}

/**
 * A ClusterRoleBinding grants permissions cluster-wide to a user or
 * set of users.
 */
export class ClusterRoleBinding extends Resource {
  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterrolebindings';

  public readonly role: clusterrole.IClusterRole;

  private readonly _subjects: Array<rb.ISubject>;

  constructor(scope: Construct, id: string, props: ClusterRoleBindingProps) {
    super(scope, id);

    this.role = props.role;

    this._subjects = new Array<rb.ISubject>();

    this.apiObject = new k8s.KubeClusterRoleBinding(this, 'Resource', {
      metadata: props.metadata,
      subjects: Lazy.any({ produce: () => this.synthesizeSubjects() }),
      roleRef: Lazy.any({ produce: () => this.synthesizeRoleRef() }),
    });
  }

  /**
   * Adds a subject to the role.
   * @param subjects The subjects to add
   */
  public addSubjects(...subjects: rb.ISubject[]): void {
    for (const subject of subjects) {
      this._subjects.push(subject);
    }
  }

  public get subjects() {
    return [...this._subjects];
  }

  private synthesizeRoleRef(): k8s.RoleRef {
    return {
      apiGroup: this.role.apiGroup,
      kind: this.role.kind,
      name: this.role.name,
    };
  }

  private synthesizeSubjects(): k8s.Subject[] {
    return this._subjects.map((subject) => filterUndefined({
      apiGroup: subject.apiGroup === 'core' ? '' : subject.apiGroup,
      kind: subject.kind,
      name: subject.name,
      namespace: subject.namespace,
    }));
  }
}
