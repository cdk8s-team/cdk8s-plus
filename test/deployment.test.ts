import { Testing, ApiObject, Duration } from 'cdk8s';
import { Node } from 'constructs';
import * as kplus from '../src';
import { DeploymentStrategy, PercentOrAbsolute, k8s } from '../src';

test('defaultChild', () => {

  const chart = Testing.chart();

  const defaultChild = Node.of(new kplus.Deployment(chart, 'Deployment')).defaultChild as ApiObject;

  expect(defaultChild.kind).toEqual('Deployment');

});

test('A label selector is automatically allocated', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment');
  deployment.addContainer({ image: 'foobar' });

  const expectedValue = 'test-Deployment-c83f5e59';
  const expectedSelector = { 'cdk8s.io/metadata.addr': expectedValue };

  // assert the k8s spec has it.
  const spec = Testing.synth(chart)[0].spec;
  expect(spec.selector.matchLabels).toEqual(expectedSelector);
  expect(spec.template.metadata?.labels).toEqual(expectedSelector);

  // assert the deployment object has it.
  expect(deployment.matchLabels).toEqual(expectedSelector);

});


test('Can be isolated', () => {

  const chart = Testing.chart();

  new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'foobar' }],
    replicas: 5,
    isolate: true,
  });

  const manifest = Testing.synth(chart);
  expect(manifest).toMatchSnapshot();

  const networkPolicy = manifest[1].spec;
  expect(networkPolicy.podSelector.matchLabels).toBeDefined;
  expect(networkPolicy.policyTypes).toEqual(['Egress', 'Ingress']);
});

test('No selector is generated if "select" is false', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    select: false,
    containers: [{ image: 'foobar' }],
  });

  // assert the k8s spec doesnt have it.
  const spec = Testing.synth(chart)[0].spec;
  expect(spec.selector.matchLabels).toBeUndefined();

  // assert the deployment object doesnt have it.
  expect(deployment.matchLabels).toEqual({});

});

test('Can select by label', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [
      {
        image: 'image',
      },
    ],
    select: false,
  });

  const expectedSelector = { foo: 'bar' };

  deployment.select(kplus.LabelSelector.of({ labels: { foo: expectedSelector.foo } }));

  // assert the k8s spec has it.
  const spec = Testing.synth(chart)[0].spec;
  expect(spec.selector.matchLabels).toEqual(expectedSelector);

  // assert the deployment object has it.
  expect(deployment.matchLabels).toEqual(expectedSelector);

});

test('Can be exposed as via ingress', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [
      {
        image: 'image',
        port: 9300,
      },
    ],
  });

  deployment.exposeViaIngress('/hello');

  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('Expose uses the correct default values', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [
      {
        image: 'image',
        port: 9300,
      },
    ],
  });

  deployment.exposeViaService();

  const spec = Testing.synth(chart)[1].spec;
  expect(spec.type).toEqual('ClusterIP');
  expect(spec.ports![0].targetPort).toEqual(9300);
  expect(spec.ports![0].port).toEqual(9300);

  const replicas = Testing.synth(chart)[0].spec.replicas;
  expect(replicas).toEqual(2);
});

test('Expose can set service and port details', () => {
  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [
      {
        image: 'image',
        port: 9500,
      },
    ],
  });

  deployment.exposeViaService({
    ports: [{ port: 9200, protocol: kplus.Protocol.UDP, targetPort: 9500 }],
    name: 'test-srv',
    serviceType: kplus.ServiceType.CLUSTER_IP,
  });

  const srv = Testing.synth(chart)[1];
  const spec = srv.spec;

  expect(srv.metadata.name).toEqual('test-srv');
  expect(spec.type).toEqual('ClusterIP');
  expect(spec.selector).toEqual({
    'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
  });
  expect(spec.ports![0].port).toEqual(9200);
  expect(spec.ports![0].targetPort).toEqual(9500);
  expect(spec.ports![0].protocol).toEqual('UDP');
});


test('Cannot be exposed if there are no containers in spec', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment');

  expect(() => deployment.exposeViaService()).toThrowError('Unable to expose deployment');
});

test('Synthesizes spec lazily', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment');
  deployment.addContainer(
    {
      image: 'image',
      port: 9300,
    },
  );

  const container = Testing.synth(chart)[0].spec.template.spec.containers[0];

  expect(container.image).toEqual('image');
  expect(container.ports[0].containerPort).toEqual(9300);

});

test('default deployment strategy', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment');
  deployment.addContainer({ image: 'image' });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.strategy).toEqual(DeploymentStrategy.rollingUpdate());
  expect(spec.strategy).toEqual({
    type: 'RollingUpdate',
    rollingUpdate: {
      maxSurge: '25%',
      maxUnavailable: '25%',
    },
  });

});

test('custom deployment strategy', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    strategy: DeploymentStrategy.recreate(),
  });
  deployment.addContainer({ image: 'image' });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.strategy).toEqual(DeploymentStrategy.recreate());
  expect(spec.strategy).toEqual({ type: 'Recreate' });

});

test('rolling update deployment strategy with a custom maxSurge and maxUnavailable', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    strategy: DeploymentStrategy.rollingUpdate({
      maxSurge: PercentOrAbsolute.percent(50),
      maxUnavailable: PercentOrAbsolute.absolute(1),
    }),
  });
  deployment.addContainer({ image: 'image' });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(spec.strategy).toEqual({
    type: 'RollingUpdate',
    rollingUpdate: {
      maxSurge: '50%',
      maxUnavailable: 1,
    },
  });
});

test('throws is maxSurge and maxUnavailable is set to zero for rolling update', () => {

  const chart = Testing.chart();

  expect(() => new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    strategy: DeploymentStrategy.rollingUpdate({ maxSurge: PercentOrAbsolute.absolute(0), maxUnavailable: PercentOrAbsolute.percent(0) }),
  })).toThrowError('\'maxSurge\' and \'maxUnavailable\' cannot be both zero');
});

test('PercentOrAbsoulte zero', () => {

  expect(PercentOrAbsolute.percent(0).isZero()).toBeTruthy();
  expect(PercentOrAbsolute.absolute(0).isZero()).toBeTruthy();
  expect(PercentOrAbsolute.percent(1).isZero()).toBeFalsy();
  expect(PercentOrAbsolute.absolute(1).isZero()).toBeFalsy();

});

test('default minReadySeconds', () => {

  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
  });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.minReady).toEqual(Duration.seconds(0));
  expect(spec.minReadySeconds).toEqual(0);

});

test('default progressDeadlineSeconds', () => {

  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
  });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.progressDeadline).toEqual(Duration.seconds(600));
  expect(spec.progressDeadlineSeconds).toEqual(600);

});

test('can configure minReadySeconds', () => {

  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    minReady: Duration.seconds(60),
  });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.minReady).toEqual(Duration.seconds(60));
  expect(spec.minReadySeconds).toEqual(60);

});

test('can configure progressDeadlineSeconds', () => {

  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    progressDeadline: Duration.seconds(60),
  });

  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;

  expect(deployment.progressDeadline).toEqual(Duration.seconds(60));
  expect(spec.progressDeadlineSeconds).toEqual(60);

});

test('throws if minReadySeconds > progressDeadlineSeconds', () => {

  const chart = Testing.chart();

  expect(() => new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    minReady: Duration.seconds(60),
    progressDeadline: Duration.seconds(30),
  })).toThrowError("'progressDeadline' (30s) must be greater than 'minReady' (60s)");

});

test('throws if minReadySeconds = progressDeadlineSeconds', () => {

  const chart = Testing.chart();

  expect(() => new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    minReady: Duration.seconds(60),
    progressDeadline: Duration.seconds(60),
  })).toThrowError("'progressDeadline' (60s) must be greater than 'minReady' (60s)");

});

test('can select with expressions', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{ image: 'image' }],
    select: false,
  });

  deployment.select(kplus.LabelSelector.of({
    expressions: [
      kplus.LabelExpression.in('foo', ['v1', 'v2']),
      kplus.LabelExpression.notIn('foo', ['v1', 'v2']),
      kplus.LabelExpression.exists('foo'),
      kplus.LabelExpression.doesNotExist('foo'),
    ],
  }));

  const expected: Set<k8s.LabelSelectorRequirement> = new Set([
    { key: 'foo', operator: 'In', values: ['v1', 'v2'] },
    { key: 'foo', operator: 'NotIn', values: ['v1', 'v2'] },
    { key: 'foo', operator: 'Exists' },
    { key: 'foo', operator: 'DoesNotExist' },
  ]);

  // assert the k8s spec has it.
  const spec: k8s.DeploymentSpec = Testing.synth(chart)[0].spec;
  expect(new Set(spec.selector.matchExpressions)).toEqual(expected);
});

describe('scheduling', () => {

  test('can tolerate tainted nodes', () => {

    const chart = Testing.chart();

    const devNodes = kplus.Node.tainted(
      kplus.NodeTaintQuery.is('key1', 'value1'),
      kplus.NodeTaintQuery.is('key2', 'value2', { effect: kplus.TaintEffect.PREFER_NO_SCHEDULE }),
      kplus.NodeTaintQuery.exists('key3'),
      kplus.NodeTaintQuery.exists('key4', { effect: kplus.TaintEffect.NO_SCHEDULE }),
      kplus.NodeTaintQuery.is('key5', 'value5', {
        effect: kplus.TaintEffect.NO_EXECUTE,
        evictAfter: Duration.hours(1),
      }),
      kplus.NodeTaintQuery.any(),
    );
    const redis = new kplus.Pod(chart, 'Redis', { containers: [{ image: 'redis' }] });
    redis.scheduling.tolerate(devNodes);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be assigned to a node by name', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', { containers: [{ image: 'redis' }] });
    redis.scheduling.assign(kplus.Node.named('node1'));

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be attracted to a node by selector - default', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', { containers: [{ image: 'redis' }] });
    redis.scheduling.attract(kplus.Node.labeled(kplus.NodeLabelQuery.is('memory', 'high')));

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be attracted to a node by selector - custom', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', {
      containers: [{ image: 'redis' }],
    });
    redis.scheduling.attract(kplus.Node.labeled(kplus.NodeLabelQuery.is('memory', 'high')), {
      weight: 1,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be co-located with a managed deployment - default', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', {
      containers: [{ image: 'redis' }],
    });
    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.colocate(redis);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be co-located with a managed deployment - custom', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', {
      containers: [{ image: 'redis' }],
    });
    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.colocate(redis, {
      topology: kplus.Topology.ZONE,
      weight: 1,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be co-located with an unmanaged deployment - default', () => {

    const chart = Testing.chart();

    const redis = kplus.Pods.select(chart, 'Redis', {
      labels: { app: 'store' },
      namespaces: kplus.Namespaces.select(chart, 'Net', { labels: { net: '1' } } ),
    });

    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.colocate(redis);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be co-located with an unmanaged deployment - custom', () => {

    const chart = Testing.chart();

    const redis = kplus.Pods.select(chart, 'Redis', {
      labels: { app: 'store' },
      namespaces: kplus.Namespaces.select(chart, 'Net', { labels: { net: '1' } } ),
    });

    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.colocate(redis, {
      topology: kplus.Topology.ZONE,
      weight: 1,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be spread - default', () => {

    const chart = Testing.chart();

    const deployment = new kplus.Deployment(chart, 'Deployment', {
      containers: [{ image: 'redis' }],
    });

    deployment.scheduling.spread();

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be spread - custom', () => {

    const chart = Testing.chart();

    const deployment = new kplus.Deployment(chart, 'Deployment', {
      containers: [{ image: 'redis' }],
    });

    deployment.scheduling.spread({ weight: 1 });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('spread set to true', () => {

    const chart = Testing.chart();

    new kplus.Deployment(chart, 'Deployment', {
      containers: [{ image: 'redis' }],
      spread: true,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();
  });

  test('can be separated from a managed deployment - default', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', {
      containers: [{ image: 'redis' }],
    });
    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.separate(redis);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be separated from a managed deployment - custom', () => {

    const chart = Testing.chart();

    const redis = new kplus.Deployment(chart, 'Redis', {
      containers: [{ image: 'redis' }],
    });
    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.separate(redis, {
      topology: kplus.Topology.ZONE,
      weight: 1,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be separated from an unmanaged deployment - default', () => {

    const chart = Testing.chart();

    const redis = kplus.Pods.select(chart, 'Redis', {
      labels: { app: 'store' },
      namespaces: kplus.Namespaces.select(chart, 'Net', { labels: { net: '1' } } ),
    });

    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.separate(redis);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('can be separated from an unmanaged deployment - custom', () => {

    const chart = Testing.chart();

    const redis = kplus.Pods.select(chart, 'Redis', {
      labels: { app: 'store' },
      namespaces: kplus.Namespaces.select(chart, 'Net', { labels: { net: '1' } } ),
    });

    const web = new kplus.Deployment(chart, 'Web', {
      containers: [{ image: 'web' }],
    });

    web.scheduling.separate(redis, {
      topology: kplus.Topology.ZONE,
      weight: 1,
    });

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

});

test('exposing via a service preserves deployment namespace', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{
      image: 'image',
      port: 9300,
    }],
    metadata: {
      namespace: 'custom',
    },
  });

  deployment.exposeViaService();
  expect(Testing.synth(chart)).toMatchSnapshot();

});

test('expose captures all container ports', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{
      image: 'image',
      ports: [
        {
          number: 8080,
          name: 'port1',
        },
        {
          number: 9090,
          name: 'port2',
        },
      ],
    }],
  });

  deployment.exposeViaService();
  expect(Testing.synth(chart)).toMatchSnapshot();

});

test('cannot expose with a port not owned by the container', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{
      image: 'image',
      ports: [
        {
          number: 8080,
        },
        {
          number: 9090,
        },
      ],
    }],
  });

  // this should fail because none of the containers in this
  // deployment expose port 2020.
  expect(() => deployment.exposeViaService({
    ports: [{ port: 2020 }],
  })).toThrowError('Unable to expose deployment');

});

test('expose via service with multiple ports throws error when names are not provided', () => {

  const chart = Testing.chart();

  const deployment = new kplus.Deployment(chart, 'Deployment', {
    containers: [{
      image: 'image',
      ports: [
        {
          number: 8080,
          name: 'port1',
        },
        {
          number: 9090,
        },
      ],
    }],
  });

  expect(() => deployment.exposeViaService()).toThrowError('Unable to expose deployment');

  const deployment1 = new kplus.Deployment(chart, 'Deployment2', {
    containers: [{
      image: 'image',
      ports: [
        {
          number: 8080,
        },
        {
          number: 9090,
        },
      ],
    }],
  });

  expect(() => deployment1.exposeViaService()).toThrowError('Unable to expose deployment');
});