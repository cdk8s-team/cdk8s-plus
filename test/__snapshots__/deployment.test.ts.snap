// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can be exposed as via ingress 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment-c898c72d",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.deployment": "test-Deployment-c83f5e59",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.deployment": "test-Deployment-c83f5e59",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [],
              "image": "image",
              "imagePullPolicy": "Always",
              "name": "main",
              "ports": Array [
                Object {
                  "containerPort": 9300,
                },
              ],
              "volumeMounts": Array [],
            },
          ],
          "volumes": Array [],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-deployment-service-c870ff98",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 9300,
          "targetPort": 9300,
        },
      ],
      "selector": Object {
        "cdk8s.deployment": "test-Deployment-c83f5e59",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "test-deployment-service-ingress-c8520013",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "service": Object {
                    "name": "test-deployment-service-c870ff98",
                    "port": Object {
                      "number": 9300,
                    },
                  },
                },
                "path": "/hello",
                "pathType": "Prefix",
              },
            ],
          },
        },
      ],
    },
  },
]
`;
